course5/homework02/homework0202.cc

```diff
@@ -0,0 +1,59 @@
+#include <iostream>
+#include <cmath>
+
+/****
+homework02, part 2
+Write a recursive function that calculates square root using Newton’s method.
+Let the function returns not only the result of the computation, but also some meta data:
+- Depth of the recursions
+- Enough data that allows you to infer the size of a stack frame of the function. There is several ways of doing it; for instance: you can record the address of some local variable in the function in the first and last call of the function, divided it by (depth-1)
+
+Answer the following:
+- which direction does the stack grows?
+- what is the impact of compilation mode (debug/release) on the size of the stack frame
+
+
+Newton's Method, from Wikipedia:
+a method for finding successively better approximations to the roots (or zeroes) of a real-valued function.
+Given a function f and an initial guess x_n at the root, then a better approximation is:
+x_{n+1} = x_{n} - f(x_{n}) / f'(x_{n})
+
+So for square roots:
+f(x) = x^2
+f'(x) = 2x
+x_{n+1} = x_{n} - (x_{n}^2 - x_{n}) / (2 * x_{n})
+***/
+
+float newton_sqrt(float input, float result, int num_iters, float** addr_of_last_on_stack)
+{
+    float arbitrarily_reassigned_input = input;
+    std::cout << "Frames remaining  : " << num_iters << std::endl;
+    std::cout << "  frame start addr: " << &arbitrarily_reassigned_input << std::endl;
```

Comment:
Oh yeah -- we're actually copying the parameter into a local variable in the new stack frame. I didn't think about it as a local variable because it wasn't initialized in the { body } itself.